test/e2e/scale.go

```diff
@@ -0,0 +1,577 @@
+	ClusterNamePrefix string
+	// BaseClusterTemplateYAML is used to create the clusters. All the clusters will use
+	// the same YAML and with only the name of the cluster changed.
+	BaseClusterTemplateYAML []byte
+	// BaseClusterName should be equal to the name of the cluster used in the BaseClusterTemplateYAML
+	// Note: It is important that the name matches the name from the BaseClusterTemplateYAML as it will
+	// replace from the template to create new clusters with new names.
+	BaseClusterName string
+
+	FailFast bool
+
+	clusterctl.ApplyClusterTemplateAndWaitInput
+}
+
+func createClustersConcurrentlyAndWait(ctx context.Context, input createClustersConcurrentlyAndWaitInput) []string {
+	Expect(ctx).NotTo(BeNil(), "ctx is required for createClustersConcurrentlyAndWait")
+	Expect(input.Concurrency).To(BeNumerically(">", 0), "Invalid argument. input.Concurrency should be greater that 0")
+	Expect(input.Namespace).NotTo(BeEmpty(), "Invalid argument. input.Namespace cannot be empty")
+	Expect(input.BaseClusterTemplateYAML).NotTo(BeNil(), "Invalid argument. input.BaseClusterTemplateYAML cannot be empty")
+	Expect(input.BaseClusterName).NotTo(BeEmpty(), "Invalid argument. input.BaseClusterName cannot be empty")
+	Expect(input.ClusterNamePrefix).NotTo(BeEmpty(), "Invalid argument. input.ClusterNamePrefix cannot be empty")
+	Expect(input.ClusterProxy).NotTo(BeNil(), "Invalid argument. input.ClusterProxy cannot be nil")
+
+	// Start a channel. This channel will be used to coordinate work with the creation workers.
+	// The channel is used to communicate the name of the cluster to be created.
+	// Adding a new name to the channel implies that a new cluster of the given names needs to be created.
+	createChan := make(chan string)
+	createWG := &sync.WaitGroup{}
+	doneChan := make(chan bool)
+	errChan := make(chan errResult)
+
+	// Start the creation workers.
+	for i := 0; i < input.Concurrency; i++ {
+		createWG.Add(1)
+		go createClusterAndWaitWorker(ctx, createChan, createWG, input.Namespace, input.BaseClusterTemplateYAML, input.BaseClusterName, input.ApplyClusterTemplateAndWaitInput, errChan)
+	}
+
+	clusterNames := make([]string, 0, input.ClusterCount)
+	// Create the workload clusters.
+	go func() {
+		for i := 1; i <= input.ClusterCount; i++ {
+			name := fmt.Sprintf("%s-%d", input.ClusterNamePrefix, i)
+			clusterNames = append(clusterNames, name)
+			createChan <- name
+		}
+		// All the clusters are requested to be created.
+		// Close the channel to shut down workers as they become unused.
+		close(createChan)
+	}()
+
+	go func() {
+		createWG.Wait()
+		close(doneChan)
+	}()
+
+	errResults := []errResult{}
+
+outer:
+	for {
+		select {
+		case err := <-errChan:
+			errResults = append(errResults, err)
+			if input.FailFast {
+				break outer
+			}
+		case <-doneChan:
+			break outer
+		}
+	}
+
+	// Clean up. All the workers are shut down.
+	// Close the error channel.
+	close(errChan)
```

Comment:
Not sure if we should close here:
```go
// The close built-in function closes a channel, which must be either
// bidirectional or send-only. It should be executed only by the sender,
// never the receiver
```

Reply:
https://stackoverflow.com/questions/24096026/in-go-what-happens-if-you-write-to-closed-channel-can-i-treat-channels-as-deter
> If you write to a closed channel, your program will panic (see http://play.golang.org/p/KU7MLrFQSx for example). You could potentially [catch this error with recover](http://golang.org/ref/spec#Handling_panics), but being in a situation where you don't know whether the channel you are writing to is open is usually a sign of a bug in the program.

Reply:
I think in general this could be fine but we have to ensure the workers are already done before we do this

I think this could roughly work like this:
* Pass a cancelable context into createClusterAndWaitWorker
* Introduce a select in createClusterAndWaitWorker to return if the context is canceled / Done
* change the errChan to a buffered channel 
* cancel the context in l.351 then wait until the doneChan is closed then empty the errChan

I think right now workers could panic in recover when we just close the channel. Happy to pair a bit to play around with it.

Reply:
I think we can de-dup createClustersConcurrentlyAndWait and deleteClustersConcurrentlyAndWait

They are mostly the same with some differences that we can pass in and some differences which are probably not intended